File: 2023/CPP/Day_1_2/Day_1_2/Day_1_2.cpp
// Day_1_2.cpp : Diese Datei enthält die Funktion "main". Hier beginnt und endet die Ausführung des Programms.
//

#include <iostream>
#include <fstream>
#include <string>
#include <map>

int calculateCalibrationSum(const std::string& filePath) {
    std::ifstream inputFile(filePath);
    if (!inputFile.is_open()) {
        std::cerr << "Error opening file: " << filePath << std::endl;
        return -1;
    }

    int sumOfCalibrations = 0;
    std::map<std::string, std::string> spelledOutDigits = {
        {"one", "1"}, {"two", "2"}, {"three", "3"}, {"four", "4"},
        {"five", "5"}, {"six", "6"}, {"seven", "7"}, {"eight", "8"}, {"nine", "9"}
    };

    std::string line;

    while (std::getline(inputFile, line)) {
        std::string currentDigit;
        std::string calibrationValue;

        for (char ch : line) {
            if (std::isalpha(ch)) {
                currentDigit += ch;
            }
            else if (std::isdigit(ch)) {
                if (!currentDigit.empty()) {
                    calibrationValue += spelledOutDigits[currentDigit];
                    currentDigit.clear();
                }
                calibrationValue += ch;
            }
        }

        if (!calibrationValue.empty()) {
            int value = std::stoi(calibrationValue);
            sumOfCalibrations += value;
        }
    }

    inputFile.close();
    return sumOfCalibrations;
}

int main() {
    std::string filePath = "d:\\Develop\\Python\\AdventOfCode\\2023\\input\\input_day1.txt";
    int result = calculateCalibrationSum(filePath);

    if (result != -1) {
        std::cout << "Sum of calibration values: " << result << std::endl;
    }

    return 0;
}


// Programm ausführen: STRG+F5 oder Menüeintrag "Debuggen" > "Starten ohne Debuggen starten"
// Programm debuggen: F5 oder "Debuggen" > Menü "Debuggen starten"

// Tipps für den Einstieg: 
//   1. Verwenden Sie das Projektmappen-Explorer-Fenster zum Hinzufügen/Verwalten von Dateien.
//   2. Verwenden Sie das Team Explorer-Fenster zum Herstellen einer Verbindung mit der Quellcodeverwaltung.
//   3. Verwenden Sie das Ausgabefenster, um die Buildausgabe und andere Nachrichten anzuzeigen.
//   4. Verwenden Sie das Fenster "Fehlerliste", um Fehler anzuzeigen.
//   5. Wechseln Sie zu "Projekt" > "Neues Element hinzufügen", um neue Codedateien zu erstellen, bzw. zu "Projekt" > "Vorhandenes Element hinzufügen", um dem Projekt vorhandene Codedateien hinzuzufügen.
//   6. Um dieses Projekt später erneut zu öffnen, wechseln Sie zu "Datei" > "Öffnen" > "Projekt", und wählen Sie die SLN-Datei aus.

File: 2023/Python/day_1_2.py
import itertools
import re

def find_first_and_last_occurrence(input_string):
    # Definiere die Muster für Ziffern und Wörter
    digit_pattern = re.compile(r'[1-9]')
    word_pattern = re.compile(r'one|two|three|four|five|six|seven|eight|nine')

    # Suche nach allen Übereinstimmungen im Eingabestring
    all_matches = re.findall(digit_pattern, input_string) + [word_to_digit(match) for match in re.findall(word_pattern, input_string)]

    if not all_matches:
        return None, None

    # Finde das erste und letzte Vorkommen
    first_occurrence = all_matches[0]
    last_occurrence = all_matches[-1]

    return first_occurrence, last_occurrence

def word_to_digit(word):
    # Übersetze Wort in Ziffer
    word_to_digit_mapping = {'one': '1', 'two': '2', 'three': '3', 'four': '4', 'five': '5', 'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'}
    return word_to_digit_mapping.get(word, word)

def calculate_calibration_sum(file_path):
    with open(file_path, 'r') as file:
        lines = file.readlines()

    sum_of_calibrations = 0

    spelled_out_digits = {
        'one': '1', 'two': '2', 'three': '3', 'four': '4', 'five': '5',
        'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'
    }

    for line in lines:
        first, last = find_first_and_last_occurrence(line)
        
        if first is None:
            first = ""

        if last is None:
            last = ""

        #if first is not None and last is not None:
        calibration_value = int(first + last)
        sum_of_calibrations += calibration_value

    return sum_of_calibrations

file_path = '../input/input_day1.txt'
result = calculate_calibration_sum(file_path)
print("Sum of calibration values:", result)
